docs/contributing/backward-compatibility-promise.md

```diff
@@ -0,0 +1,144 @@
+New [translation messages](/docs/introduction/translations.md) may be added or have its translation changed in any release.
+However, they may be removed or have their message ID changed only in `MAJOR` releases.
+
+When changing the message (eg. because of a typo) in a `MINOR` or `PATCH` release, only its translation may be changed.
+In such instances, it is preferred to keep using the original message ID to ensure backward compatibility with existing user-defined translation.
+This might be unintuitive for contributors because we use English text as message IDs.
+See an example of fixing a typo in an English translation message:
+
+Initially, there is only the message ID in the `messages.en.po` (translation doesn't have to be filled):
+```diff
+  msgid "Exaple translation"
+  msgstr ""
+```
+
+In a `PATCH` or `MINOR` version, the original message ID may not be removed:
+```diff
+  msgid "Exaple translation"
+- msgstr ""
++ msgstr "Example translation"
+```
+
+In a `MAJOR` version, the original message ID with the typo may be removed:
+```diff
+- msgid "Exaple translation"
+- msgstr "Example translation"
++ msgid "Example translation"
++ msgstr ""
+```
+
+### Routing
+New routes may be added in any release.
+Existing routes should not be modified, except for adding a new [optional placeholder](https://symfony.com/doc/3.4/routing/optional_placeholders.html).
+
+Route names may be changed only in a `MAJOR` release.
+
+### Docker Configuration and Orchestration Manifests
+Configuration of containers and orchestration is mostly in the `project-base` repository, which means that [it's excluded form the BC promise](#project-base-repository).
+
+The changes should always be described in [upgrade notes](/UPGRADE.md) (in the *Infrastructure* section).
+
+### Twig
+Twig blocks, functions, filters and the templates themselves can be removed or renamed only in a `MAJOR` release.
+
+Changes of Twig functions and filters in `MINOR` and `PATCH` releases must be backward-compatible.
+This means only a new optional argument or a support for new data type of existing argument may be added.
+
+### LESS / CSS
+Classes can be renamed only in a `MAJOR` release.
+
+Changes of the visual style of the administration are not considered a BC break and may be introduced in any version.
+
+The style of the front-end is in the `project-base` repository, which means that [it's excluded form the BC promise](#project-base-repository).
+
+### Javascript
+Javascript code should adhere to similar rules as the PHP code ([except in project-base repository](#project-base-repository)).
+
+Behavior of elements with classes prefixed with `.js-` should change only in a `MAJOR` release.
+New keys in a `options` map or a new expected data attribute may be added in a `MINOR` or `PATCH` release only if it has a default value.
+Adding a new optional argument to a method is considered a BC break (as it the method might be overridden by the user with the new argument already in use).
```

Comment:
```suggestion
Adding a new optional argument to a method is considered as a BC break (as it the method might be overridden by the user with the new argument already in use) and is not allowed in `MINOR` or `PATCH` release.
```

Reply:
`a method is considered a BC break` is OK imo (https://ludwig.guru/s/it+can+be+considered+a)

I've added the what-it-means part at the end, though